BaseLib/ConfigTree-impl.h

```diff
@@ -50,15 +52,37 @@ getConfParam(std::string const& param, T const& default_value) const
     return default_value;
 }
 
-template<typename T>
-boost::optional<T>
-ConfigTree::
-getConfParamOptional(std::string const& param) const
+template <typename T>
+boost::optional<T> ConfigTree::getConfParamOptional(
+    std::string const& param) const
 {
     checkUnique(param);
 
+    return getConfParamOptionalImpl(param, static_cast<T*>(nullptr));
+}
+
+template <typename T>
+boost::optional<T> ConfigTree::getConfParamOptionalImpl(
```

Comment:
I think @endJunction's solution is the best one. The problem is that you cannot have partial specializations if templated __methods__. So, e.g., if you declare a template method
```
    template<typename T>
    void a();
```
and a specialization
```
    template<typename T>
    void a<vector<T>>();
```
you will get a compilation error like
`non-type partial specialization »a<std::vector<T, std::allocator<_CharT> > >« is not allowed`.